Switch CoreTelephony.framework to use delay-init linking

ddkilzer opened the pull request:
#### 3e53dd942f96f2a13275beadcdcf0ea2f517b58f
<pre>
Switch CoreTelephony.framework to use delay-init linking
<a href="https://bugs.webkit.org/show_bug.cgi?id=271936">https://bugs.webkit.org/show_bug.cgi?id=271936</a>
&lt;<a href="https://rdar.apple.com/125661316">rdar://125661316</a>&gt;

Reviewed by Elliott Williams and Wenson Hsieh.

* Source/WTF/wtf/PlatformHave.h:
(HAVE_DELAY_INIT_LINKING): Add.
- Define macro for platforms that support delay-init linking.
* Source/WebCore/PAL/pal/cocoa/CoreTelephonySoftLink.h:
* Source/WebCore/PAL/pal/cocoa/CoreTelephonySoftLink.mm:
- Comment out soft-linking macros when using delay-init linking.
* Source/WebKit/Configurations/WebKit.xcconfig:
(OTHER_LDFLAGS):
(OTHER_LDFLAGS_DELAY_INIT): Add.
- Add linker flag for delay-init of CoreTelephony.framework.
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.mm:
(WebKit::shouldAllowAutoFillForCellularIdentifiers):
- Update source code for use with delay-init linking.
* Tools/TestWebKitAPI/Configurations/TestWebKitAPILibrary.xcconfig:
(OTHER_LDFLAGS): Add.
(OTHER_LDFLAGS_DELAY_INIT): Add.
- Add linker flag for delay-init of CoreTelephony.framework.
* Tools/TestWebKitAPI/Tests/ios/KeyboardInputTestsIOS.mm:
(TestWebKitAPI::TEST(KeyboardInputTests, DeviceEIDAndIMEIAutoFill)):
- Update source code for use with delay-init linking.

Canonical link: <a href="https://commits.webkit.org/276947@main">https://commits.webkit.org/276947@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d8a20629c876bbe90aa0333834625b64aacdf6b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46005 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25134 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48590 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48676 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42045 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48312 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29445 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22535 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37624 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46583 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22218 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39685 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18815 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19615 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40781 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4049 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/39238 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42345 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41132 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50460 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/45479 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21001 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17481 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44782 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22301 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43673 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22660 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/52630 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6447 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21995 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10751 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.mm:
```diff
@@ -77,7 +77,11 @@ bool shouldAllowAutoFillForCellularIdentifiers(const URL& topURL)
     }
 
+#if HAVE(DELAY_INIT_LINKING)
+    static NeverDestroyed cachedClient = adoptNS([[CoreTelephonyClient alloc] initWithQueue:dispatch_get_main_queue()]);
```
If CoreTelephony.framework is missing from an SDK, this is now a build-time failure. If it's missing at runtime, it's a launch failure. Previously it would have built regardless, and only crashed in this function (which may not be compiled at all due to the HAVE guard). Is that safe?

ddkilzer replied:
I can confirm the `CoreTelephonyClient` Objective-C class and `-initWithQueue:` method are available in `CoreTelephony.framework` in iOS 17 and later—they have been around for a long time.

Checking on SDKs and images.

ddkilzer replied:
> Checking on SDKs and images.

Running some internal commands to check SDKs and images, I don't think the `CoreTelephonyClient` Objective-C class will cause any build-time or launch-time issues.

I also checked with a manager on the team, and they weren't aware of any places where `CoreTelephony.framework` might not be available.

ddkilzer replied:
Add @whsieh to the reviewers list in case he has anything to add about the "intent" when soft-linking `CoreTelephony.framework`.

Commit:
Switch CoreTelephony.framework to use delay-init linking
https://bugs.webkit.org/show_bug.cgi?id=271936
<rdar://125661316>

Reviewed by Elliott Williams and Wenson Hsieh.

* Source/WTF/wtf/PlatformHave.h:
(HAVE_DELAY_INIT_LINKING): Add.
- Define macro for platforms that support delay-init linking.
* Source/WebCore/PAL/pal/cocoa/CoreTelephonySoftLink.h:
* Source/WebCore/PAL/pal/cocoa/CoreTelephonySoftLink.mm:
- Comment out soft-linking macros when using delay-init linking.
* Source/WebKit/Configurations/WebKit.xcconfig:
(OTHER_LDFLAGS):
(OTHER_LDFLAGS_DELAY_INIT): Add.
- Add linker flag for delay-init of CoreTelephony.framework.
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.mm:
(WebKit::shouldAllowAutoFillForCellularIdentifiers):
- Update source code for use with delay-init linking.
* Tools/TestWebKitAPI/Configurations/TestWebKitAPILibrary.xcconfig:
(OTHER_LDFLAGS): Add.
(OTHER_LDFLAGS_DELAY_INIT): Add.
- Add linker flag for delay-init of CoreTelephony.framework.
* Tools/TestWebKitAPI/Tests/ios/KeyboardInputTestsIOS.mm:
(TestWebKitAPI::TEST(KeyboardInputTests, DeviceEIDAndIMEIAutoFill)):
- Update source code for use with delay-init linking.

Canonical link: https://commits.webkit.org/276947@main
